newlib/libc/tinystdio/atod_engine.c

```diff
@@ -37,13 +37,16 @@ __atod_engine(uint64_t u64, int exp)
 {
     double flt = u64;
 
-    while (exp < 0) {
-        flt *= 1e-1;
-        exp++;
-    }
-    while (exp > 0) {
-        flt *= 1e1;
-        exp--;
+    if (exp < 0) {
+        do {
+            flt *= 1e-1;
+            exp++;
+        } while (exp < 0);
+    } else if (exp > 0) {
+        do {
+            flt *= 1e1;
+            exp--;
+        } while (exp > 0);
```

Comment:
Looks similar to the bug hit in dirname. I decided to skip the test in `run-arm` for that case.
https://gcc.gnu.org/bugzilla/show_bug.cgi?id=110791

Reply:
Oh, now I remember why I didn't -- this bug doesn't just cause the test to fail, it causes it to infinite loop, which we can't fix with hacks in run-arm. I've made the work-around simpler *and* made it specific to GCC 12.2.1 with -Os.

Reply:
oh, and applied the same fix to `__atof_engine`

Reply:
I like the solution with changing the optimization mode. It is much nicer, than ignoring the bug by skipping the test or hacking nice code.